backends/graphs/graphs.cpp

```diff
@@ -0,0 +1,128 @@
+#include "frontends/p4/evaluator/evaluator.h"
+#include "frontends/p4/frontend.h"
+
+#include "config.h"
+#ifdef HAVE_LIBBOOST_GRAPH
+#include "controls.h"
+#endif
+
+namespace graphs {
+
+class MidEnd : public PassManager {
+ public:
+    P4::ReferenceMap    refMap;
+    P4::TypeMap         typeMap;
+    IR::ToplevelBlock   *toplevel = nullptr;
+
+    explicit MidEnd(CompilerOptions& options);
+    IR::ToplevelBlock* process(const IR::P4Program *&program) {
+        program = program->apply(*this);
+        return toplevel;
+    }
+};
+
+MidEnd::MidEnd(CompilerOptions& options) {
+    bool isv1 = options.langVersion == CompilerOptions::FrontendVersion::P4_14;
+    refMap.setIsV1(isv1);
+    auto evaluator = new P4::EvaluatorPass(&refMap, &typeMap);
+    setName("MidEnd");
+
+    addPasses({
+        evaluator,
+        new VisitFunctor([this, evaluator]() { toplevel = evaluator->getToplevelBlock(); }),
+    });
+}
+
+class Options : public CompilerOptions {
+ public:
+    cstring graphsDir{"."};
+    Options() {
+        registerOption("--graphs-dir", "dir",
+                       [this](const char* arg) { graphsDir = arg; return true; },
+                       "Use this directory to dump graphs in dot format "
+                       "(default is current working directory)\n");
+     }
+};
+
+}  // namespace graphs
+
+int main(int argc, char *const argv[]) {
+    setup_gc_logging();
+    setup_signals();
+
+    graphs::Options options;
+    options.langVersion = CompilerOptions::FrontendVersion::P4_16;
+    options.compilerVersion = "0.0.5";
+
+    if (options.process(argc, argv) != nullptr)
+        options.setInputFile();
+    if (::errorCount() > 0)
+        return 1;
+
+    auto hook = options.getDebugHook();
+
+    auto program = P4::parseP4File(options);
+    if (program == nullptr || ::errorCount() > 0)
+        return 1;
+
+    try {
+        P4::FrontEnd fe;
+        fe.addDebugHook(hook);
+        program = fe.run(options, program);
+    } catch (const Util::P4CExceptionBase &bug) {
+        std::cerr << bug.what() << std::endl;
+        return 1;
+    }
+    if (program == nullptr || ::errorCount() > 0)
+        return 1;
+
+    graphs::MidEnd midEnd(options);
+    midEnd.addDebugHook(hook);
+    const IR::ToplevelBlock *top = nullptr;
+    try {
+        top = midEnd.process(program);
+        if (options.dumpJsonFile)
+            JSONGenerator(*openFile(options.dumpJsonFile, true)) << program << std::endl;
+    } catch (const Util::P4CExceptionBase &bug) {
+        std::cerr << bug.what() << std::endl;
+        return 1;
+    }
+    if (::errorCount() > 0)
+        return 1;
+
+#ifdef HAVE_LIBBOOST_GRAPH
+    LOG1("Generating graphs under " << options.graphsDir);
+    LOG1("Generating control graphs");
+    graphs::ControlGraphs cgen(&midEnd.refMap, &midEnd.typeMap, options.graphsDir);
+    top->getMain()->apply(cgen);
+#else
+    ::error("This backend requires the boost graph headers, which could not be found");
```

Comment:
This looks like a runtime error, but the problem happened when building the back-end.
Either the error message should be clearer, or the build should fail.